agasc/scripts/update_obs_status.py

```diff
@@ -0,0 +1,235 @@
+        (obs['obsid'], agasc_id): {
+            'ok': obs['ok'],
+            'comments': obs['comments']
+        }
+        for obs in status['obs'] for agasc_id in obs['agasc_id']
+    }
+
+    return status
+
+
+def parse_obs_status_args(filename=None, bad_star=[], bad_star_source=None,
+                          obs=None, status=None, comments='', agasc_id=None,
+                          **_
+                          ):
+    """
+    Combine obs/bad-star status from file and from arguments.
+
+    The arguments could be passed from and ArgumentParser doing::
+
+        parse_obs_status_args(vars(args))
+
+    :param filename: str
+    :param bad_star: int
+    :param bad_star_source: int
+    :param obs: int
+    :param status: int
+    :param comments: str
+    :param agasc_id: int
+    :return:
+    """
+    obs_status_override = {}
+    bad_star = (list(np.atleast_1d(bad_star)))
+    bad = {}
+
+    if bad_star and bad_star_source is None:
+        raise RuntimeError('If you specify bad_star, you must specify bad_star_source')
+
+    if filename is not None:
+        status_file = parse_obs_status_file(filename)
+        bad = status_file['bad']
+        obs_status_override = status_file['obs']
+
+    if obs is not None and status is not None:
+        if agasc_id is None:
+            from agasc.supplement.magnitudes import star_obs_catalogs
+            agasc_ids = list(
+                sorted(star_obs_catalogs.STARS_OBS[star_obs_catalogs.STARS_OBS['obsid'] == obs][
+                       'agasc_id'])
+            )
+        else:
+            agasc_ids = sorted(list(np.atleast_1d(agasc_id)))
+
+        for agasc_id in agasc_ids:
+            obs_status_override[(obs, agasc_id)] = {'ok': status, 'comments': comments}
+
+    for bs in bad_star:
+        if bs in bad and bad[bs] != bad_star_source:
+            raise RuntimeError('name collision: conflicting bad_star in file and in args')
+        bad[bs] = bad_star_source
+
+    return {'obs': obs_status_override, 'bad': bad}
+
+
+def update_obs_status(filename, obs_status_override, dry_run=False):
+    """
+    Update the magnitude table of the AGASC supplement.
+
+    :param filename:
+    :param obs_status_override: dict.
+        Dictionary with OK flag for specific observations.
+        Keys are (OBSID, AGASC ID) pairs, values are dictionaries like
+        {'ok': True, 'comments': 'some comment'}
+    :param dry_run: bool
+    :return:
+    """
+    logger = logging.getLogger('agasc.supplement')
+
+    if not pathlib.Path(filename).exists():
+        raise FileExistsError(f'AGASC supplement file does not exist: {filename}')
+
+    if not obs_status_override:
+        logger.info('Nothing to update')
+        return
+
+    try:
+        obs_status = table.Table.read(filename, format='hdf5', path='obs')
+    except OSError as e:
+        if str(e) == 'Path obs does not exist':
+            obs_status = {}
+        else:
+            raise
+
+    obs_status = {
```

Comment:
This all works, but I think it could be much simpler. Start by making a dict of keys that map to the table index.
```
idxs = {key:idx for idx, key in enumerate(zip(obs_status['obsid'], obs_status['agasc_id']))}
```
Then (assuming that `obs_status` is always an appropriately initialized `Table` at this point, perhaps zero-length)
```
for key, status in obs_status_override.items():
    row_vals = (key[0], key[1], status['ok'], status['comments'])
    if key in idxs:
        obs_status[idxs[key]] = row_vals  # Update in place
    else:
        obs_status.add_row(row_vals)  # New row
```